group-coordinator/src/main/java/org/apache/kafka/coordinator/group/generic/GenericGroup.java

```diff
@@ -898,6 +900,52 @@ public void createGroupTombstoneRecords(List<Record> records) {
         records.add(RecordHelpers.newGroupMetadataTombstoneRecord(groupId()));
     }
 
+    @Override
+    public boolean isGroupEmpty() {
+        return isInState(EMPTY);
+    }
+
+    /**
+     * Return the offset expiration condition to be used for this group. This is based on several factors
+     * such as the group state, the protocol type, and the GroupMetadata record version.
+     *
+     * See {@link org.apache.kafka.coordinator.group.OffsetMetadataManager.OffsetExpirationCondition}
+     *
+     * @return The offset expiration condition for the group or Empty of no such condition exists.
+     */
+    @Override
+    public Optional<OffsetMetadataManager.OffsetExpirationCondition> offsetExpirationCondition() {
+        if (protocolType.isPresent()) {
+            if (isInState(EMPTY)) {
+                // No consumer exists in the group =>
+                // - If current state timestamp exists and retention period has passed since group became Empty,
+                //   expire all offsets with no pending offset commit;
+                // - If there is no current state timestamp (old group metadata schema) and retention period has passed
+                //   since the last commit timestamp, expire the offset
+                return Optional.of(
+                    (offsetAndMetadata, currentTimestamp, offsetsRetentionMs) -> OffsetMetadataManager.isExpiredOffset(
+                        currentTimestamp,
+                        currentStateTimestamp.orElse(offsetAndMetadata.commitTimestampMs),
+                        offsetAndMetadata.expireTimestampMs,
+                        offsetsRetentionMs
+                    )
```

Comment:
The dependency between the classes feels a bit weird here. The OffsetMetadataManager calls offsetExpirationCondition which returns an OffsetExpirationCondition which calls the OffsetMetadataManager. I was not expecting this when I suggested it.

We could perhaps have an implementation of the OffsetExpirationCondition which contains the logic of `OffsetMetadataManager.isExpiredOffset` and takes a function to extract the timestamp from the record. Here we could just instantiate the new class with the correct way to get the timestamp. Would it make sense?

Reply:
Thanks for the suggestion! I moved OffsetExpirationCondition to a separate file along with a new OffsetExpirationConditionImpl class. Let me know how this is.